File: Decorator Pattern/1. Minimal/Python/main.py
from concrete_component import ConcreteComponent
from concrete_decorator_a import ConcreteDecoratorA
from concrete_decorator_b import ConcreteDecoratorB

def main():
    print("Decorator Pattern in Python")
    print("---------------------------")
    print()
    
    # Creating a concrete component.
    component = ConcreteComponent()
    print("Concrete Component:")
    component.operation()

    print()

    # Decorating the concrete component with the concrete decorator a.
    decorated_component_a = ConcreteDecoratorA(component)
    print("Decorated Component A:")
    decorated_component_a.operation()

    print()

    # Decorating the concrete component with the concrete decorator b.
    decorated_component_b = ConcreteDecoratorB(component)
    print("Decorated Component B:")
    decorated_component_b.operation()

    print()

    # Decorating the concrete component with both concrete decorators nested.
    decorated_component_ab = ConcreteDecoratorB(ConcreteDecoratorA(component))
    print("Decorated Component AB:")
    decorated_component_ab.operation()

if __name__ == "__main__":
    main()

File: Decorator Pattern/1. Minimal/Java/Main.java
public class Main {
    public static void main(String[] args) {
        System.out.println("Decorator Pattern in Java");
        System.out.println("-------------------------");
        System.out.println();
        
        // Creating a concrete component.
        AbstractComponent component = new ConcreteComponent();
        System.out.println("Concrete Component:");
        component.operation();

        System.out.println();

        // Decorating the concrete component with the concrete decorator a.
        AbstractComponent decoratedCompontentA = new ConcreteDecoraterA(component);
        System.out.println("Decorated Component A:");
        decoratedCompontentA.operation();

        System.out.println();

        // Decorating the concrete component with the concrete decorator b.
        AbstractComponent decoratedComponentB = new ConcreteDecoratorB(component);
        System.out.println("Decorated Component B:");
        decoratedComponentB.operation();

        System.out.println();

        // Decorating the concrete component with both concrete decorators nested.
        AbstractComponent decoratedComponentAB = new ConcreteDecoratorB(new ConcreteDecoraterA(component));
        System.out.println("Decorated Component AB:");
        decoratedComponentAB.operation();
    }
}
